[FG:InPlacePodVerticalScaling] Introduce  /resize subresource to request pod resource resizing

AnishShah opened the pull request:
#### What type of PR is this?


/kind feature
/kind api-change


#### What this PR does / why we need it:

This PR is a continuation to @iholder101's PR (https://github.com/kubernetes/kubernetes/pull/127320) that originally introduced /resize subresource and @mouuii's PR (https://github.com/kubernetes/kubernetes/pull/124887) that added support for pod resize in LimitRanger admission plugin. 

Introduce  /resize subresource to request pod resource resizing

#### Which issue(s) this PR fixes:

Fixes #109553
Fixes #124855

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
A new /resize subresource was added to request pod resource resizing. Update your k8s client code to utilize the /resize subresource for Pod resizing operations.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources
```

AnishShah committed:
introduce resize REST object

This commit introduces a basic REST object for resize subresource and
adds it to the pod storage.

AnishShah committed:
validate resize request

We validate resize request by ensuring that pod QoS is unchanged and
only cpu and memory resources and resize policies are mutated.

AnishShah committed:
drop unrelated changes for /resize request

Since resize request takes the full pod object as the request type, drop
any unrelated changes. Only container resources and resize policy should
be validated.

AnishShah committed:
e2e: Use resize subresource in e2e tests

AnishShah committed:
api: generate client code for resize subresource

AnishShah committed:
integration tests for pod resize

AnishShah commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2
/test pull-kubernetes-node-kubelet-serial-podresize

AnishShah commented:
/assign @tallclair

tallclair commented:
/assign @jpbetz

tallclair commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -281,6 +272,58 @@ func (podEphemeralContainersStrategy) WarningsOnUpdate(ctx context.Context, obj,
+
+// dropNonPodResizeUpdates discards all changes except for pod.Spec.Containers[*].Resources,ResizePolicy and certain metadata
+func dropNonPodResizeUpdates(newPod, oldPod *api.Pod) *api.Pod {
+	pod := oldPod.DeepCopy()
+	pod.Name = newPod.Name
+	pod.Namespace = newPod.Namespace
+	pod.ResourceVersion = newPod.ResourceVersion
+	pod.UID = newPod.UID
```
nit: make this a helper?

tallclair commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -281,6 +272,58 @@ func (podEphemeralContainersStrategy) WarningsOnUpdate(ctx context.Context, obj,
+		}
+		pod.Spec.Containers[idx].Resources = ctr.Resources
+		pod.Spec.Containers[idx].ResizePolicy = ctr.ResizePolicy
+	}
```
I think doing this in `PrepareForUpdate` means any other changes will get silently dropped. That's probably OK for status & metadata, but for spec it might be better to fail in validation.

AnishShah replied:
Not only for `status` and `metadata`, we have the same behavior for `ephemeralcontainers` subresource. Hence, I decided to silently drop. WDYT?

jpbetz replied:
I was also wondering the approach here.  I saw `"pod resize may not change fields other than cpu and memory"` in validation but also saw `dropNonPodResizeUpdates` and wasn't clear which approach was being taken here.

FWIW, we've always wiped status when updating spec and vis-versa, and the `ephemeralcontainers` precedence is compelling.  I think I'm in favor of wiping non-updated fields, purely for consistency reasons.  But curious what others think.

jpbetz replied:
The part I'm confused about is:  Are changes to "fields other than cpu and memory" being dropped and ignored, or are such changes being reported as errors?  Or both?

thockin replied:
If they are dropped, then we can kind of pretend they don't exist, and users can specify ONLY the changed fields, right?  That seems better than validating it and risking missing things.  Sort of treat it as unknown fields?

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5488,6 +5442,65 @@ func ValidatePodEphemeralContainersUpdate(newPod, oldPod *core.Pod, opts PodVali
+	// newPod.Spec.Containers[].Resources are allowed.
+	specPath := field.NewPath("spec")
+	if qos.GetPodQOS(oldPod) != qos.ComputePodQOS(newPod) {
+		allErrs = append(allErrs, field.Invalid(specPath, newPod.Status.QOSClass, "Pod QoS is immutable"))
```
```suggestion
		allErrs = append(allErrs, field.Invalid(specPath, newPod.Status.QOSClass, "Pod QOS Class must not change"))
```

tallclair reviewed (commented):
Nice. My only open question is whether we need to do something with managed fields, but I'll let @jpbetz comment on that.

Also, please update the release note to mention that non-resize pod updates can no longer mutate resources.

jpbetz commented on test/e2e/common/node/pod_resize.go:
```diff
@@ -894,8 +894,8 @@ func doPodResizeTests(f *framework.Framework) {
 				ginkgo.By(fmt.Sprintf("patching pod for %s", opStr))
-				patchedPod, pErr = f.ClientSet.CoreV1().Pods(newPod.Namespace).Patch(context.TODO(), newPod.Name,
-					types.StrategicMergePatchType, []byte(patchString), metav1.PatchOptions{})
+				patchedPod, pErr = f.ClientSet.CoreV1().Pods(newPod.Namespace).Patch(ctx, newPod.Name,
```
Curious if there is a reason to test with `Patch` instead of `Resize` here.  I see some integration tests use `Resize`, and I see value to testing via all possible paths (`Resize`, `Patch` and `Update`)..

jpbetz replied:
Gotcha. A few other things to consider here:  Will developers be using these tests are references for best practices?  If so, what approach to we want to encourage?  Also, do we have sufficient test coverage for `Resize`?  Would converting some of these tests help improve coverage?

I mean these as genuine questions, I haven't spent the time to assess what the answers are, but wanted to bring these considerations up.

AnishShah replied:
> Will developers be using these tests are references for best practices? If so, what approach to we want to encourage?

I think `Patch` request is always better since it takes only required changes and avoids resource conflicts.

> Also, do we have sufficient test coverage for Resize? Would converting some of these tests help improve coverage?

e2e tests in this file are more to test kubelet behavior rather than k8s client behavior. integration tests might be better fit to test `Resize` and `Patch`. I have added [basic integration tests](https://github.com/kubernetes/kubernetes/pull/128266/files#diff-4bc3f052b24d529f60a80616320bed5cea00e447f65e5dee1511706d2f595d09R785) for both of them. The reason for this is because we have [unit tests](https://github.com/kubernetes/kubernetes/pull/128266/files#diff-7ee023b4401073292170227985c8ac98015db0c1980657c277a7cea011c24830R24821) covering all possible validation cases.

jpbetz replied:
> > Will developers be using these tests are references for best practices? If so, what approach to we want to encourage?
> 
> I think `Patch` request is always better since it takes only required changes and avoids resource conflicts.

Makes sense, I've [added a separate suggestion](https://github.com/kubernetes/kubernetes/pull/128266/files#r1815206956) based on this line of thinking.

jpbetz commented on test/e2e/common/node/pod_resize.go:
```diff
@@ -894,8 +894,8 @@ func doPodResizeTests(f *framework.Framework) {
 
```
Worth adding a test that shows using RBAC to only allow subject to update /resize, but not the entire pod, and then verify that the subject is able to update via resize but not by writes to the root pod reosurce?

AnishShah replied:
Done. But I added integration tests instead of e2e tests.

jpbetz commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -281,6 +272,58 @@ func (podEphemeralContainersStrategy) WarningsOnUpdate(ctx context.Context, obj,
+		if !ok {
+			continue
+		}
+		pod.Spec.Containers[idx].Resources = ctr.Resources
+		pod.Spec.Containers[idx].ResizePolicy = ctr.ResizePolicy
```
Preserve `InitContainers` too?  I read that [they were supported in KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#subresource).

The KEP also mentions `.spec.resizePolicy`.  should that field be preserved also?

jpbetz replied:
Gotcha.

How about `.spec.resizePolicy`?

tallclair replied:
Yeah, we don't have plans to add a pod-level resizePolicy. Sorry for the confusion.

jpbetz commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -281,6 +272,58 @@ func (podEphemeralContainersStrategy) WarningsOnUpdate(ctx context.Context, obj,
+	oldPod := old.(*api.Pod)
+
+	*newPod = *dropNonPodResizeUpdates(newPod, oldPod)
+	podutil.MarkPodProposedForResize(oldPod, newPod)
```
Should some of the tests verify that the `.status.resize` is being written to as expected?

AnishShah committed:
test: add unit tests to verify the .status.resize field

AnishShah committed:
refactor logic to override pod fields

AnishShah committed:
fix error message for pod resize validation failure

AnishShah committed:
integration: RBAC tests for /resize request

The test confirms that the subject can successfully
resize the Pod resources but not the entire pod.

AnishShah commented:
/retest

jpbetz commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4867,6 +4867,7 @@ type PodStatusResult struct {
 
 // +genclient
 // +genclient:method=UpdateEphemeralContainers,verb=update,subresource=ephemeralcontainers
+// +genclient:method=Resize,verb=update,subresource=resize
```
```suggestion
// +genclient:method=UpdateResize,verb=update,subresource=resize
// +genclient:method=ApplyResize,verb=apply,subresource=resize
// +genclient:method=PatchResize,verb=patch,subresource=resize
```

Since patch and apply are more useful for this feature, and this would be more consistent with what is generated for ephemeralcontainers and status.

liggitt replied:
I don't think we need applypatch, can use the generic patch and pass in a subresource

for apply, there's questions about how well apply will work (see the ResetFields comment on the strategy)

thockin replied:
Do we have a sense of whether subresources supporting patch really matters?  I guess if you are passing the whole pod here it matters, but for something like /scale - why wouldn't everyone just use PUT?

AnishShah replied:
> I don't think we need applypatch, can use the generic patch and pass in a subresource

What is the difference between the generic patch and the patch suggested here ?

> for something like /scale - why wouldn't everyone just use PUT?

I might be wrong here. But can there be resource version mismatch incase of scale subresource?

jpbetz replied:
My understanding from a quick chat during an API review session with @liggitt  is that we don't need the "PatchResize" generation and most of our other subresources do not provide (since it can be performed on the root resource by passing a "resize" into the subresource vararg.  This same logic is also be valid for Apply.  So the only change I think I'd still make here is renaming the function to `UpdateResize` for consistency with other subresource update function names.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5166,45 +5156,12 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		return allErrs
 	}
 
-	if qos.GetPodQOS(oldPod) != qos.ComputePodQOS(newPod) {
```
relaxing this validation unconditionally on pod spec update is surprising to me... I thought this PR was _removing_ the ability to change resourcey stuff via the normal pod spec update endpoint

also, xref https://github.com/kubernetes/kubernetes/pull/127744 which is attempting to *tighten* QOS validation on pod status update

SergeyKanzhelev replied:
I believe we agreed to keep this check and disallow the change of QoS. @tallclair to keep me honest on the latest in all those discussions we had.

tallclair replied:
+1 to what Anish said. This check is redundant with the check that resources are immutable, for regular pod-update. For the resize subresources, the check is still there.

liggitt commented on pkg/registry/core/rest/storage_core.go:
```diff
@@ -242,6 +242,9 @@ func (p *legacyProvider) NewRESTStorage(apiResourceConfigSource serverstorage.AP
 		}
 		storage[resource+"/ephemeralcontainers"] = podStorage.EphemeralContainers
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+			storage[resource+"/resize"] = podStorage.Resize
```
Making this a subresource means that existing admission policies that intercept pod create that have opinions about pod resources (or mutating admission that is auto-adjusting resources) will be bypassed (xref https://github.com/kubernetes/kubernetes/issues/113594)

That *may* be ok if:
1. the permission to write to this new subresource is not granted to anyone other than superusers by default (similar to pods/ephemeralcontainers)
2. we document *very* clearly that before a cluster admin grants this permission, they ensure validating / mutating admission policies are updated to also register for and handle `UPDATE pods/resize` requests

tallclair replied:
Good points. TBH I'm still unsure of whether the subresource is the right decision.

Regarding general admission, this is still a problem without a dedicated subresource, right? Webhooks that deal with resources might not be looking at pod updates since they were previously immutable. Most of the motivation for the subresource was that it's easier to restrict access to just resize, but the flip side is also true (easier to prevent resize permission).

+1 for updating quota. Limitrange is racing with https://github.com/kubernetes/kubernetes/pull/124887

liggitt commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -281,6 +268,65 @@ func (podEphemeralContainersStrategy) WarningsOnUpdate(ctx context.Context, obj,
+}
+
+// ResizeStrategy wraps and exports the used podStrategy for the storage package.
+var ResizeStrategy = podResizeStrategy{Strategy}
```
this has to implement GetResetFields() and enumerate *every* field path that gets reset for server-side apply to work properly

```
spec.... (every individual parent path to fields that are untouched by requests to this endpoint)
metadata
status
```

tallclair replied:
Seems like more reason to not reset the spec fields, and instead fail it in validation?

jpbetz replied:
I checked and it looks like /ephemeralcontainers fails validation.  I had assumed it reset fields, but I was wrong.

Maybe that consistency with /ephemeralcontainers + problems with GetResetFields() is enough reason to validate here?

I'll defer to @thockin and @liggitt for API review decision here.

jpbetz replied:
> I am not super versed in how SSA actually works, but it seems like enumerating every field is going to be a nightmare. Can't we somehow say "all except..." ?

I've opened https://github.com/kubernetes/kubernetes/issues/128345 to track this, since it seems like something we need to address.  If the outcome of the "should we passing a whole pod type here" ends up being a *yes*, please ping the issue and SIG API Machinery can bump the prioritization.

AnishShah replied:
Done.

liggitt commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -281,6 +268,65 @@ func (podEphemeralContainersStrategy) WarningsOnUpdate(ctx context.Context, obj,
+func dropNonPodResizeUpdates(newPod, oldPod *api.Pod) *api.Pod {
+	pod := mungePod(newPod, oldPod)
+
+	oldCtrToIndex := make(map[string]int)
```
does something ensure that the number of containers does not change, and that `old[i].name == new[i].name`?

AnishShah replied:
I don't believe so. That's why, I added this logic.

jpbetz commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5166,45 +5156,12 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
-			lim := mungeCpuMemResources(container.Resources.Limits, oldPod.Spec.Containers[ix].Resources.Limits)
-			req := mungeCpuMemResources(container.Resources.Requests, oldPod.Spec.Containers[ix].Resources.Requests)
-			container.Resources = core.ResourceRequirements{Limits: lim, Requests: req}
-		}
```
How would a kubectl user perform a resize?

BenTheElder commented:
/triage accepted

AnishShah committed:
support for resize subresource in ResourceQuota admission.

AnishShah committed:
pod resize support in LimitRanger admission plugin

AnishShah committed:
update codegen

AnishShah commented:
/retest

jpbetz commented on pkg/quota/v1/evaluator/core/pods.go:
```diff
@@ -155,14 +155,11 @@ func (p *podEvaluator) GroupResource() schema.GroupResource {
 
 // Handles returns true if the evaluator should handle the specified attributes.
 func (p *podEvaluator) Handles(a admission.Attributes) bool {
-	op := a.GetOperation()
```
I'm concerned this won't actually work due to:

https://github.com/kubernetes/kubernetes/blob/5b1fffa3e40b812e81ede244f671c90e3428e2ec/pkg/kubeapiserver/admission/exclusion/resources.go#L30

Can we add a test that proves me wrong? (adding /resize to that include list if needed?)

jpbetz replied:
Actually... per our offline conversation.. I think this included list is only for ValidatingAdmissionPolicy. We still need to add resize to the list, but quota should work.

jpbetz commented on pkg/quota/v1/evaluator/core/resource_claims_test.go:
```diff
@@ -170,3 +172,52 @@ func TestResourceClaimEvaluatorMatchingResources(t *testing.T) {
 	}
 }
+
+func TestResourceClaimEvaluatorHandles(t *testing.T) {
```
Huge thanks for adding test coverage for these evaluators.

thockin commented:
First note: the release not could be a bit more precise about what it means for end-users.  Do they need to change how they do something?

AnishShah commented:
> First note: the release not could be a bit more precise about what it means for end-users. Do they need to change how they do something?

I updated the release note. They basically need to update their k8s client code to use the new subresource.

thockin reviewed (commented):
I'm a little outside my comfort-zone WRT the machinery, but I am really questioning why passing the whole pod is correct.

AnishShah committed:
test: cleanup validation tests

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5488,6 +5442,68 @@ func ValidatePodEphemeralContainersUpdate(newPod, oldPod *core.Pod, opts PodVali
+func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) field.ErrorList {
+	// Part 1: Validate newPod's spec and updates to metadata
+	fldPath := field.NewPath("metadata")
+	allErrs := ValidateObjectMetaUpdate(&newPod.ObjectMeta, &oldPod.ObjectMeta, fldPath)
```
As long as we're munging the metadata in PrepareForUpdate, I think it's moot. I think it's safe to just do a `DeepEqual` here.

AnishShah replied:
I don't think we're munging the metadata for resize or other subresources like status, ephemeralcontainers. 

I was able to add a pod label with this command:

```
kubectl --context kind-kind patch pod pause -p '{"metadata":{"labels":{"environment":"production"}}}' --subresource=status
```

liggitt replied:
The starting expectation is that subresources prevent (either by ignoring or forbidding) changes to fields outside the specific ones they allow addressing.

`pods/status` allows metadata mutation for compatibility reasons, but that was an oversight early in Kubernetes.

`pods/ephemeralcontainers` definitely *does* prevent mutation of fields outside `spec.ephemeralContainers`:

https://github.com/kubernetes/kubernetes/blob/78fed9dbcdc9ed10727dedaf09a95931482c7554/pkg/registry/core/pod/strategy.go#L252-L261

AnishShah replied:
Marking this thread as resolved. To answer Tim's question in the first comment, we don't allow mutating any metadata with resize subresource.

thockin replied:
Can we memorialize such discussions in the code by asserting no change and tests?

AnishShah replied:
Done. Added unit tests to ensure metadata cannot be updated via resize subresource.

AnishShah committed:
unit tests to ensure pod metadata cannot be updated during resize.

AnishShah committed:
client-go: rename Resize to UpdateResize

AnishShah committed:
better variable names

AnishShah committed:
implement GetResetFieldsFilter

GetResetFieldsFilter returns a set of fields filter reset
by pod resize strategy. This is needed to make server-side apply
work correctly.

jpbetz commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -281,6 +268,76 @@ func (podEphemeralContainersStrategy) WarningsOnUpdate(ctx context.Context, obj,
+// and should not be modified by the user.
+func (podResizeStrategy) GetResetFieldsFilter() map[fieldpath.APIVersion]fieldpath.Filter {
+	return map[fieldpath.APIVersion]fieldpath.Filter{
+		"v1": fieldpath.NewIncludeMatcherFilter(
+			fieldpath.MakePrefixMatcherOrDie("spec", "containers", fieldpath.MatchAnyPathElement(), "resources"),
+			fieldpath.MakePrefixMatcherOrDie("spec", "containers", fieldpath.MatchAnyPathElement(), "resizePolicy"),
+		),
+	}
```
If you care to optimize, you could construct this at podResizeStrategy initialization time and reuse.

thockin reviewed (commented):
sending notes now, I did not finish my readthru yet

SergeyKanzhelev commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -281,6 +268,76 @@ func (podEphemeralContainersStrategy) WarningsOnUpdate(ctx context.Context, obj,
+		if !ok {
+			continue
+		}
+		pod.Spec.Containers[idx].Resources = ctr.Resources
```
pls put a TODO for sidecar support to coordinate with this PR: https://github.com/kubernetes/kubernetes/pull/128367

SergeyKanzhelev commented on plugin/pkg/admission/limitranger/admission.go:
```diff
@@ -417,6 +417,12 @@ func (d *DefaultLimitRangerActions) ValidateLimit(limitRange *corev1.LimitRange,
 // SupportsAttributes ignores all calls that do not deal with pod resources or storage requests (PVCs).
 // Also ignores any call that has a subresource defined.
 func (d *DefaultLimitRangerActions) SupportsAttributes(a admission.Attributes) bool {
+	// Handle the special case for in-place pod vertical scaling
```
why do we call it "special case" in comments? Maybe explain what so special about it

AnishShah replied:
Done.

SergeyKanzhelev commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:
```diff
@@ -155,10 +155,6 @@ func (a *QuotaAdmission) ValidateInitialization() error {
 
 // Validate makes admission decisions while enforcing quota
 func (a *QuotaAdmission) Validate(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) (err error) {
-	// ignore all operations that correspond to sub-resource actions
-	if attr.GetSubresource() != "" {
-		return nil
-	}
```
how bad this may be? Can code below have some assumptions about subresource already being checked and get broken by this change? Is there any perf implications? I wonder if we should only allow resize subresource here

SergeyKanzhelev replied:
I don't know this logic well enough to reason about it. It just looks scary that we introduce a large change here without any safeguards in case we missed anything.

@liggitt @jpbetz who know this code best

AnishShah replied:
> It just looks scary that we introduce a large change here without any safeguards in case we missed anything.

Yeah, I had a similar concern and hence I added unit tests to avoid any side-effects.

esotsal replied:
We will need to update e2e tests to get more confidence, I believe. Will it be a separate PR for this ?

esotsal replied:
Oh, I saw it is already done in this PR ! Good work @AnishShah !

AnishShah commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -281,6 +268,76 @@ func (podEphemeralContainersStrategy) WarningsOnUpdate(ctx context.Context, obj,
+func dropNonPodResizeUpdates(newPod, oldPod *api.Pod) *api.Pod {
+	pod := dropPodUpdates(newPod, oldPod)
+
+	oldCtrToIndex := make(map[string]int)
```
Done.

AnishShah commented on pkg/api/pod/testing/make.go:
```diff
@@ -297,3 +298,34 @@ func SetContainerRestartPolicy(policy api.ContainerRestartPolicy) TweakContainer
 }
+
+func MakePodStatus(tweaks ...TweakPodStatus) api.PodStatus {
+	ps := api.PodStatus{}
```
> Does an empty PodStatus pass validation? 

I did not understand this question. By default the `Pod` object created in this file has `Status` field set to nil. But we provide [`SetStatus`](https://github.com/kubernetes/kubernetes/blob/1b119d4f30f64f2edf49d9b8be093367d8af76ab/pkg/api/pod/testing/make.go#L217) func to set it. I introduced `MakePodStatus` to introduce a way to build pod status easily.

> The goal here should be there's no boilerplate code that callers have to sdo to pass tests -- every tweak is directly related to that specific test case

I think this is true here too. We can use `MakePodStatus` (and tweaks) only if a test case requires pod status validation.

AnishShah committed:
apply feedback

AnishShah commented:
/retest

thockin reviewed (commented):
API LGTM

thockin commented:
/approve

/hold for kubelet and other reviewers

AnishShah commented:
/retest

tallclair commented:
/priority important-soon

tallclair commented:
/milestone v1.32

tallclair commented on test/e2e/node/pod_resize.go:
```diff
@@ -37,122 +37,180 @@ import (
+
+			ginkgo.By("creating pods")
+			podClient := e2epod.NewPodClient(f)
+			newPod1 := podClient.CreateSync(ctx, testPod1)
+			newPod2 := podClient.CreateSync(ctx, testPod2)
```
nit: use `CreateBatch` to create these in parallel

tallclair commented on test/e2e/node/pod_resize.go:
```diff
@@ -37,122 +37,180 @@ import (
+	}
+
+	for _, tc := range testcases {
+		ginkgo.It(tc.name, func(ctx context.Context) {
```
maybe execute each test case in a separate namespace? Then you don't need to worry about cleanup or test cases interacting with each other.

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5531,6 +5485,68 @@ func ValidatePodEphemeralContainersUpdate(newPod, oldPod *core.Pod, opts PodVali
+func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) field.ErrorList {
+	// Part 1: Validate newPod's spec and updates to metadata
+	fldPath := field.NewPath("metadata")
+	allErrs := ValidateObjectMetaUpdate(&newPod.ObjectMeta, &oldPod.ObjectMeta, fldPath)
```
Should we change this to be:
```suggestion
	allErrs := ValidateImmutableField(&newPod.ObjectMeta, &oldPod.ObjectMeta, fldPath)
```

thockin replied:
I think validation should not assume anything about the calling context.

tallclair replied:
> ValidateImmutableField naming would be incorrect as this function is being used in pod updates too.

What do you mean? Isn't this function (`ValidatePodResize`) only called for the resize subresource strategy?

> I think validation should not assume anything about the calling context.

Is that an argument for not checking metadata at all, not assuming that metadata changes have been reset?

tallclair replied:
To clarify my original comment: there's already a function `ValidateImmutableField`, I was just suggesting we call that one instead.

https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L360-L362

thockin replied:
> Is that an argument for not checking metadata at all, not assuming that metadata changes have been reset?

The latter - if we are validating that only X and Y are changed, then that should include asserting that metadata has not changed.

AnishShah replied:
I initially thought you meant renaming `ValidateObjectMetaUpdate` to `ValidateImmutableField`. But I realized now that you meant calling an existing func `ValidateImmutableField`. I added this block again.

tallclair reviewed (commented):
Couple little things, overall LGTM

AnishShah commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2
/test pull-kubernetes-node-kubelet-serial-podresize

tallclair commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -286,6 +273,85 @@ func (podEphemeralContainersStrategy) WarningsOnUpdate(ctx context.Context, obj,
+// dropPodUpdates drops any changes in the pod.
+func dropPodUpdates(newPod, oldPod *api.Pod) *api.Pod {
+	pod := oldPod.DeepCopy()
+	pod.Name = newPod.Name
+	pod.Namespace = newPod.Namespace
+	pod.ResourceVersion = newPod.ResourceVersion
+	pod.UID = newPod.UID
```
If we're not doing any metadata validation, there's not any reason to copy these, right? Actually, it seems dangerous to copy them without metadata validation.

AnishShah replied:
added metadata validation back again.

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5531,6 +5485,68 @@ func ValidatePodEphemeralContainersUpdate(newPod, oldPod *core.Pod, opts PodVali
+	fldPath := field.NewPath("metadata")
+	allErrs := ValidateImmutableField(&newPod.ObjectMeta, &oldPod.ObjectMeta, fldPath)
+	allErrs = append(allErrs, validatePodMetadataAndSpec(newPod, opts)...)
+	allErrs = append(allErrs, ValidatePodSpecificAnnotationUpdates(newPod, oldPod, fldPath.Child("annotations"), opts)...)
```
I think this is redundant with the immutability check now.

AnishShah replied:
Done.

tallclair commented:
/lgtm

This is blocking a lot of follow-up work, so I'm going to remove the hold. Please clean up the redundant validation in a follow-up.

/hold cancel

AnishShah commented:
/retest

AnishShah committed:
remove redundant validation check for pod resize

AnishShah commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2
/test pull-kubernetes-node-kubelet-serial-podresize

tallclair commented on test/e2e/node/pod_resize.go:
```diff
@@ -37,107 +37,167 @@ import (
+
+	for _, tc := range testcases {
+		ginkgo.It(tc.name, func(ctx context.Context) {
+			ns, err := f.CreateNamespace(ctx, tc.name, nil)
```
I think it might be easier to create a new framework, rather than a namespace.

AnishShah replied:
Done.

tallclair commented on test/e2e/node/pod_resize.go:
```diff
@@ -37,107 +37,167 @@ import (
+			testPod2 = e2epod.MustMixinRestrictedPodSecurity(testPod2)
+
+			ginkgo.By("creating pods")
+			podClient := e2epod.NewPodClient(f)
```
NewPodClient uses the framework's namespace, not the new namespace you created.

AnishShah replied:
Done.

tallclair commented on test/e2e/node/pod_resize.go:
```diff
@@ -37,107 +37,167 @@ import (
+			e2epod.VerifyPodResources(patchedPodExceedCPU, expected)
+			framework.ExpectNoError(e2epod.VerifyPodStatusResources(patchedPodExceedMemory, expected))
+
+			ginkgo.By("deleting pods")
+			delErr1 := e2epod.DeletePodWithWait(ctx, f.ClientSet, newPods[0])
+			framework.ExpectNoError(delErr1, "failed to delete pod %s", newPods[0].Name)
+			delErr2 := e2epod.DeletePodWithWait(ctx, f.ClientSet, newPods[1])
+			framework.ExpectNoError(delErr2, "failed to delete pod %s", newPods[1].Name)
```
I think you remove this since the tests run in separate namespaces (should speed up the tests a bit)

AnishShah replied:
Done.

AnishShah committed:
apply feedback

AnishShah commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2
/test pull-kubernetes-node-kubelet-serial-podresize

tallclair commented:
/lgtm

AnishShah commented:
/retest

AnishShah committed:
Fix unit tests

AnishShah committed:
update codegen

AnishShah commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2
/test pull-kubernetes-node-kubelet-serial-podresize

AnishShah commented:
/retest

AnishShah commented:
/retest

tallclair commented:
/lgtm

AnishShah commented:
/retest

aojea commented on test/e2e/node/pod_resize.go:
```diff
@@ -316,13 +369,13 @@ var _ = SIGDescribe(framework.WithSerial(), "Pod InPlace Resize Container (sched
 })
 
 var _ = SIGDescribe("Pod InPlace Resize Container", feature.InPlacePodVerticalScaling, func() {
-	f := framework.NewDefaultFramework("pod-resize-tests")
 	ginkgo.BeforeEach(func(ctx context.Context) {
+		f := framework.NewDefaultFramework("pod-resize-tests")
```
@pohly do you think this can be the cause of https://github.com/kubernetes/kubernetes/issues/128600#issuecomment-2459907264

aojea replied:
It was https://github.com/kubernetes/kubernetes/issues/128607, confirmed

```
_output/bin/e2e.test --ginkgo.focus="Pod.InPlace.Resize.Container" --context kind-kind --kubeconfig=/usr/local/google/home/aojea/
.kube/config
  I1106 15:11:00.867730 2960849 test_context.go:564] The --provider flag is not set. Continuing as if --provider=skeleton had been used.
  I1106 15:11:00.867866 2960849 e2e.go:109] Starting e2e run "9452694d-8cff-4453-a16f-58957fbfc75a" on Ginkgo node 1
Running Suite: Kubernetes e2e suite - /usr/local/google/home/aojea/src/kubernetes
=================================================================================
Random Seed: 1730905860 - will randomize all specs

Will run 36 of 6598 specs
SSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
• [FAILED] [4.087 seconds]
[sig-node] [Serial] Pod InPlace Resize Container (scheduler-focused) [Feature:InPlacePodVerticalScaling] [It] pod-resize-scheduler-tests [sig-node, Serial, Feature:InPlacePodVerticalScaling]
k8s.io/kubernetes/test/e2e/node/pod_resize.go:197

  Timeline >>
  STEP: Creating a kubernetes client @ 11/06/24 15:11:01.095
  I1106 15:11:01.095575 2960849 util.go:502] >>> kubeConfig: /usr/local/google/home/aojea/.kube/config
  I1106 15:11:01.102592 2960849 util.go:511] >>> kubeContext: kind-kind
  STEP: Building a namespace api object, basename pod-resize-scheduler-tests @ 11/06/24 15:11:01.102
  STEP: Waiting for a default service account to be provisioned in namespace @ 11/06/24 15:11:01.112
  STEP: Waiting for kube-root-ca.crt to be provisioned in namespace @ 11/06/24 15:11:01.114
  I1106 15:11:01.118597 2960849 pod_resize.go:202] Found 1 schedulable nodes
  STEP: Find node CPU resources available for allocation! @ 11/06/24 15:11:01.118
  I1106 15:11:01.120278 2960849 pod_resize.go:218] Found 10 pods on node 'kind-control-plane'
  I1106 15:11:01.120291 2960849 pod_resize.go:230] Node 'kind-control-plane': NodeAllocatable MilliCPUs = 48000m. MilliCPUs currently available to allocate = 47050m.
  I1106 15:11:01.120296 2960849 pod_resize.go:242] TEST1: testPod1 initial CPU request is '23525m'
  I1106 15:11:01.120299 2960849 pod_resize.go:243] TEST1: testPod2 initial CPU request is '47050m'
  I1106 15:11:01.120302 2960849 pod_resize.go:244] TEST1: testPod2 resized CPU request is '11762m'
  STEP: TEST1: Create pod 'testpod1' that fits the node 'kind-control-plane' @ 11/06/24 15:11:01.12
  STEP: TEST1: Create pod 'testpod2' that won't fit node 'kind-control-plane' with pod 'testpod1' on it @ 11/06/24 15:11:03.13
  STEP: TEST1: Resize pod 'testpod2' to fit in node 'kind-control-plane' @ 11/06/24 15:11:05.138
  I1106 15:11:05.139391 2960849 pod_resize.go:292] Unexpected error: failed to patch pod for resize: 
      <*errors.StatusError | 0xc00098fc20>: 
      the server could not find the requested resource
      {
          ErrStatus: 
              code: 404
              details: {}
              message: the server could not find the requested resource
              metadata: {}
              reason: NotFound
              status: Failure,
      }
  [FAILED] in [It] - k8s.io/kubernetes/test/e2e/node/pod_resize.go:292 @ 11/06/24 15:11:05.139
  I1106 15:11:05.139588 2960849 helper.go:124] Waiting up to 7m0s for all (but 0) nodes to be ready
  STEP: dump namespace information after failure @ 11/06/24 15:11:05.141
  STEP: Collecting events from namespace "pod-resize-scheduler-tests-3255". @ 11/06/24 15:11:05.141
  STEP: Found 5 events. @ 11/06/24 15:11:05.142
  I1106 15:11:05.142257 2960849 dump.go:53] At 2024-11-06 15:11:01 +0000 UTC - event for testpod1: {default-scheduler } Scheduled: Successfully assigned pod-resize-scheduler-tests-3255/testpod1 to kind-control-plane
  I1106 15:11:05.142267 2960849 dump.go:53] At 2024-11-06 15:11:01 +0000 UTC - event for testpod1: {kubelet kind-control-plane} Pulled: Container image "registry.k8s.io/e2e-test-images/busybox:1.36.1-1" already present on machine
  I1106 15:11:05.142274 2960849 dump.go:53] At 2024-11-06 15:11:01 +0000 UTC - event for testpod1: {kubelet kind-control-plane} Created: Created container: c1
  I1106 15:11:05.142280 2960849 dump.go:53] At 2024-11-06 15:11:01 +0000 UTC - event for testpod1: {kubelet kind-control-plane} Started: Started container c1
  I1106 15:11:05.142285 2960849 dump.go:53] At 2024-11-06 15:11:03 +0000 UTC - event for testpod2: {default-scheduler } FailedScheduling: 0/1 nodes are available: 1 Insufficient cpu. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod.
  I1106 15:11:05.143577 2960849 resource.go:168] POD       NODE                PHASE    GRACE  CONDITIONS
  I1106 15:11:05.143606 2960849 resource.go:175] testpod1  kind-control-plane  Running         [{PodReadyToStartContainers True 0001-01-01 00:00:00 +0000 UTC 2024-11-06 15:11:02 +0000 UTC  } {Initialized True 0001-01-01 00:00:00 +0000 UTC 2024-11-06 15:11:01 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2024-11-06 15:11:02 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2024-11-06 15:11:02 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2024-11-06 15:11:01 +0000 UTC  }]
  I1106 15:11:05.143615 2960849 resource.go:175] testpod2                      Pending         [{PodScheduled False 0001-01-01 00:00:00 +0000 UTC 2024-11-06 15:11:03 +0000 UTC Unschedulable 0/1 nodes are available: 1 Insufficient cpu. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod.}]
  I1106 15:11:05.143620 2960849 resource.go:178] 
  I1106 15:11:05.144721 2960849 dump.go:109] 
  Logging node info for node kind-control-plane
  I1106 15:11:05.146004 2960849 dump.go:114] Node Info: &Node{ObjectMeta:{kind-control-plane    4b17f8b3-75b1-4ffe-9f70-596d40fe3f62 707305 0 2024-10-31 09:24:35 +0000 UTC <nil> <nil> map[beta.kubernetes.io/arch:amd64 beta.kubernetes.io/os:linux kubernetes.io/arch:amd64 kubernetes.io/hostname:kind-control-plane kubernetes.io/os:linux node-role.kubernetes.io/control-plane:] map[kubeadm.alpha.kubernetes.io/cri-socket:unix:///run/containerd/containerd.sock node.alpha.kubernetes.io/ttl:0 volumes.kubernetes.io/controller-managed-attach-detach:true] [] [] [{kubelet Update v1 2024-10-31 09:24:35 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{".":{},"f:volumes.kubernetes.io/controller-managed-attach-detach":{}},"f:labels":{".":{},"f:beta.kubernetes.io/arch":{},"f:beta.kubernetes.io/os":{},"f:kubernetes.io/arch":{},"f:kubernetes.io/hostname":{},"f:kubernetes.io/os":{}}},"f:spec":{"f:providerID":{}}} } {kubeadm Update v1 2024-10-31 09:24:36 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{"f:kubeadm.alpha.kubernetes.io/cri-socket":{}},"f:labels":{"f:node-role.kubernetes.io/control-plane":{}}}} } {kube-controller-manager Update v1 2024-10-31 09:24:43 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{"f:node.alpha.kubernetes.io/ttl":{}}},"f:spec":{"f:podCIDR":{},"f:podCIDRs":{".":{},"v:\"10.244.0.0/24\"":{}}}} } {kubelet Update v1 2024-11-06 15:06:58 +0000 UTC FieldsV1 {"f:status":{"f:conditions":{"k:{\"type\":\"DiskPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"MemoryPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"PIDPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"Ready\"}":{"f:lastHeartbeatTime":{},"f:lastTransitionTime":{},"f:message":{},"f:reason":{},"f:status":{}}},"f:images":{}}} status}]},Spec:NodeSpec{PodCIDR:10.244.0.0/24,DoNotUseExternalID:,ProviderID:kind://docker/kind/kind-control-plane,Unschedulable:false,Taints:[]Taint{},ConfigSource:nil,PodCIDRs:[10.244.0.0/24],},Status:NodeStatus{Capacity:ResourceList{cpu: {{48 0} {<nil>} 48 DecimalSI},ephemeral-storage: {{2153496870912 0} {<nil>}  BinarySI},hugepages-1Gi: {{0 0} {<nil>} 0 DecimalSI},hugepages-2Mi: {{0 0} {<nil>} 0 DecimalSI},memory: {{126605991936 0} {<nil>}  BinarySI},pods: {{110 0} {<nil>} 110 DecimalSI},},Allocatable:ResourceList{cpu: {{48 0} {<nil>} 48 DecimalSI},ephemeral-storage: {{2153496870912 0} {<nil>}  BinarySI},hugepages-1Gi: {{0 0} {<nil>} 0 DecimalSI},hugepages-2Mi: {{0 0} {<nil>} 0 DecimalSI},memory: {{126605991936 0} {<nil>}  BinarySI},pods: {{110 0} {<nil>} 110 DecimalSI},},Phase:,Conditions:[]NodeCondition{NodeCondition{Type:MemoryPressure,Status:False,LastHeartbeatTime:2024-11-06 15:06:58 +0000 UTC,LastTransitionTime:2024-10-31 09:24:35 +0000 UTC,Reason:KubeletHasSufficientMemory,Message:kubelet has sufficient memory available,},NodeCondition{Type:DiskPressure,Status:False,LastHeartbeatTime:2024-11-06 15:06:58 +0000 UTC,LastTransitionTime:2024-10-31 09:24:35 +0000 UTC,Reason:KubeletHasNoDiskPressure,Message:kubelet has no disk pressure,},NodeCondition{Type:PIDPressure,Status:False,LastHeartbeatTime:2024-11-06 15:06:58 +0000 UTC,LastTransitionTime:2024-10-31 09:24:35 +0000 UTC,Reason:KubeletHasSufficientPID,Message:kubelet has sufficient PID available,},NodeCondition{Type:Ready,Status:True,LastHeartbeatTime:2024-11-06 15:06:58 +0000 UTC,LastTransitionTime:2024-10-31 09:24:59 +0000 UTC,Reason:KubeletReady,Message:kubelet is posting ready status,},},Addresses:[]NodeAddress{NodeAddress{Type:InternalIP,Address:192.168.8.3,},NodeAddress{Type:Hostname,Address:kind-control-plane,},},DaemonEndpoints:NodeDaemonEndpoints{KubeletEndpoint:DaemonEndpoint{Port:10250,},},NodeInfo:NodeSystemInfo{MachineID:d7116c595bc346dc80ebd143e1574a80,SystemUUID:17ff3b61-28d1-4786-971b-1b18c6ba6740,BootID:141a11fe-6289-42df-b6a7-f40eb76ba33d,KernelVersion:6.9.10-1rodete5-amd64,OSImage:Debian GNU/Linux 12 (bookworm),ContainerRuntimeVersion:containerd://1.7.18,KubeletVersion:v1.32.0-alpha.2.328+cf159912e25fbd,KubeProxyVersion:v1.32.0-alpha.2.328+cf159912e25fbd,OperatingSystem:linux,Architecture:amd64,},Images:[]ContainerImage{ContainerImage{Names:[docker.io/library/import-2024-10-31@sha256:6bfbd010f2673ebc331aa4e10241ebcf648378bf523eb0fd915401a51534d2e1 registry.k8s.io/kube-apiserver-amd64:v1.32.0-alpha.2.328_cf159912e25fbd registry.k8s.io/kube-apiserver:v1.32.0-alpha.2.328_cf159912e25fbd],SizeBytes:95494129,},ContainerImage{Names:[docker.io/library/import-2024-10-31@sha256:ff641a1ca4e9ba239f64f26643e8dc69a2efbdc1b991d73ab4325c143ffce95a registry.k8s.io/kube-proxy-amd64:test registry.k8s.io/kube-proxy-amd64:v1.32.0-alpha.2.328_c4102eb7359925],SizeBytes:93927610,},ContainerImage{Names:[docker.io/library/import-2024-10-31@sha256:d0ff8bf9897902b922cf5ba82e8fa60b2f3d0bb8b6fce32a2687625d044978c4 registry.k8s.io/kube-proxy-amd64:v1.32.0-alpha.2.328_cf159912e25fbd registry.k8s.io/kube-proxy:v1.32.0-alpha.2.328_cf159912e25fbd],SizeBytes:93927098,},ContainerImage{Names:[docker.io/library/import-2024-10-31@sha256:9101f22c11b10ba4a56d26afff3af6a3070ef6179aa0556fd2ef6836b78c7c2c registry.k8s.io/kube-controller-manager-amd64:v1.32.0-alpha.2.328_cf159912e25fbd registry.k8s.io/kube-controller-manager:v1.32.0-alpha.2.328_cf159912e25fbd],SizeBytes:89480520,},ContainerImage{Names:[docker.io/library/import-2024-10-31@sha256:ff230f6561be5c3c4026ad103f5c80f62a0897ebb296efbfe1715752f229e547 registry.k8s.io/kube-scheduler-amd64:v1.32.0-alpha.2.328_cf159912e25fbd registry.k8s.io/kube-scheduler:v1.32.0-alpha.2.328_cf159912e25fbd],SizeBytes:69128008,},ContainerImage{Names:[docker.io/library/httpd@sha256:bbea29057f25d9543e6a96a8e3cc7c7c937206d20eab2323f478fdb2469d536d docker.io/library/httpd:2],SizeBytes:59389653,},ContainerImage{Names:[registry.k8s.io/etcd:3.5.16-0],SizeBytes:57680541,},ContainerImage{Names:[docker.io/kindest/kindnetd:v20241007-36f62932],SizeBytes:38600298,},ContainerImage{Names:[docker.io/kindest/local-path-provisioner:v20240813-c6f155d6],SizeBytes:19430244,},ContainerImage{Names:[registry.k8s.io/coredns/coredns:v1.11.3],SizeBytes:18562039,},ContainerImage{Names:[docker.io/kindest/local-path-helper:v20230510-486859a6],SizeBytes:3052318,},ContainerImage{Names:[registry.k8s.io/e2e-test-images/busybox@sha256:a9155b13325b2abef48e71de77bb8ac015412a566829f621d06bfae5c699b1b9 registry.k8s.io/e2e-test-images/busybox:1.36.1-1],SizeBytes:2223659,},ContainerImage{Names:[registry.k8s.io/pause:3.10],SizeBytes:320368,},},VolumesInUse:[],VolumesAttached:[]AttachedVolume{},Config:nil,RuntimeHandlers:[]NodeRuntimeHandler{},Features:nil,},}
  I1106 15:11:05.146016 2960849 dump.go:116] 
  Logging kubelet events for node kind-control-plane
  I1106 15:11:05.147145 2960849 dump.go:121] 
  Logging pods the kubelet thinks are on node kind-control-plane
  I1106 15:11:05.155983 2960849 dump.go:128] kube-system/etcd-kind-control-plane started at 2024-10-31 09:24:37 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156004 2960849 dump.go:134]    Container etcd ready: true, restart count 0
  I1106 15:11:05.156010 2960849 dump.go:128] kube-system/kube-scheduler-kind-control-plane started at 2024-10-31 09:24:37 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156016 2960849 dump.go:134]    Container kube-scheduler ready: true, restart count 0
  I1106 15:11:05.156022 2960849 dump.go:128] default/pod started at 2024-11-01 20:22:54 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156026 2960849 dump.go:134]    Container pod ready: false, restart count 1350
  I1106 15:11:05.156031 2960849 dump.go:128] kube-system/kube-proxy-spgv9 started at 2024-10-31 09:45:29 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156037 2960849 dump.go:134]    Container kube-proxy ready: true, restart count 0
  I1106 15:11:05.156042 2960849 dump.go:128] kube-system/kube-controller-manager-kind-control-plane started at 2024-10-31 09:24:37 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156048 2960849 dump.go:134]    Container kube-controller-manager ready: true, restart count 0
  I1106 15:11:05.156053 2960849 dump.go:128] kube-system/kube-apiserver-kind-control-plane started at 2024-10-31 09:24:37 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156058 2960849 dump.go:134]    Container kube-apiserver ready: true, restart count 0
  I1106 15:11:05.156063 2960849 dump.go:128] kube-system/kindnet-ngv9r started at 2024-10-31 09:24:43 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156067 2960849 dump.go:134]    Container kindnet-cni ready: true, restart count 0
  I1106 15:11:05.156071 2960849 dump.go:128] local-path-storage/local-path-provisioner-57c5987fd4-g5jv6 started at 2024-10-31 09:24:59 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156076 2960849 dump.go:134]    Container local-path-provisioner ready: true, restart count 0
  I1106 15:11:05.156081 2960849 dump.go:128] kube-system/coredns-7c65d6cfc9-2j8sw started at 2024-10-31 09:24:59 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156085 2960849 dump.go:134]    Container coredns ready: true, restart count 0
  I1106 15:11:05.156091 2960849 dump.go:128] kube-system/coredns-7c65d6cfc9-8gfxr started at 2024-10-31 09:24:59 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156095 2960849 dump.go:134]    Container coredns ready: true, restart count 0
  I1106 15:11:05.156100 2960849 dump.go:128] pod-resize-scheduler-tests-3255/testpod1 started at 2024-11-06 15:11:01 +0000 UTC (0+1 container statuses recorded)
  I1106 15:11:05.156104 2960849 dump.go:134]    Container c1 ready: true, restart count 0
  I1106 15:11:05.179476 2960849 kubelet_metrics.go:206] 
  Latency metrics for node kind-control-plane
  STEP: Destroying namespace "pod-resize-scheduler-tests-3255" for this suite. @ 11/06/24 15:11:05.179
  << Timeline

  [FAILED] failed to patch pod for resize: the server could not find the requested resource
  In [It] at: k8s.io/kubernetes/test/e2e/node/pod_resize.go:292 @ 11/06/24 15:11:05.139
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSGinkgo detected an issue with your spec structure
set up framework | framework.go:200
  It looks like you are trying to add a [BeforeEach] node
  to the Ginkgo spec tree in a leaf node after the specs started running.

  To enable randomization and parallelization Ginkgo requires the spec tree
  to be fully constructed up front.  In practice, this means that you can
  only create nodes like [BeforeEach] at the top-level or within the
  body of a Describe, Context, or When.

  Learn more at:
  http://onsi.github.io/ginkgo/#mental-model-how-ginkgo-traverses-the-spec-hierarchy
```

sftim commented:
Retrospective feedback on the changelog; this could have been
```diff
-A new /resize subresource was added to request pod resource resizing. Update your k8s client code to utilize the /resize subresource for Pod resizing operations.
+Added a new `resize` subresource to Pods; you can use this to request changes to resource requests or limits.
```
